### PR TITLE
WholeBody CT Segmentation (104 classes) | SegResNet | TotalSegmentator

### PR DESCRIPTION
**Pull request by tangy5**
Add whole body 104 tissues segmentation bundle. 

- [x] Inference pipeline
- [x] Training pipeline
- [x] Benchmarking
- [x] Performance Figures
- [x] Add Hardware requirements after benchmarking, this is a huge model, a minimum 32G GPU is required. Add more details.
- [x] Other checks
- [x] Utilities
- 
options

### CODE REVIEW

**Comment by tangy5** [comment]
Thanks, most basic config files are done, I'm benchmarking and training using A100 GPUs. 
I need to get performance comparison, train loss, validation dice figures. 
Hopefully we will get a finalized model trained with A100 GPUs. 
Will notice for review.

**Comment by tangy5** [comment]
Hi @yiheng-wang-nv , @Nic-Ma , this bundle is finished. All components are developed and evaluated. Please see details in the doc. 
It would be great that you could give a look and see whether any part needs modification, especially formats, styles. 

In additon, the CI checks failed on "code-format-check", caused by other files. 

Thank you so much.

**Comment by tangy5** [comment]
Hi @SachidanandAlle, this bundle is done, it would be great if you could give a review. I tested it with MONAI Label. 

I will create a PR in MONAI Label repo to support "options" for models versions and inference options such as "sw_batch_size", "sw_overlap". 

Thanks

**Comment by yiheng-wang-nv** [comment]
Thanks @tangy5 , the code format check issue will be resolved soon.

File: models/wholeBody_ct_segmentation/configs/metadata.json [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        "0.1.0": "complete the model package",
+        "0.0.1": "initialize the model package structure"
+    },
+    "monai_version": "1.1.0",
+    "pytorch_version": "1.13.0",
+    "numpy_version": "1.21.2",
+    "optional_packages_version": {
+        "nibabel": "4.0.1",
+        "pytorch-ignite": "0.4.9"
+    },
```
Comment (yiheng-wang-nv) [inline models/wholeBody_ct_segmentation/configs/metadata.json:14]: Hi @tangy5 , according to the latest change related to #189 , could you please also add a "name" tag here?
the name here will be the display name in the webpage

File: models/wholeBody_ct_segmentation/docs/README.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+#### Override the `train` config and `evaluate` config to execute multi-GPU evaluation
+
+```
+torchrun --standalone --nnodes=1 --nproc_per_node=2 -m monai.bundle run evaluating --meta_file configs/metadata.json --config_file "['configs/train.json','configs/evaluate.json','configs/multi_gpu_evaluate.json']" --logging_file configs/logging.conf
+```
+
+#### Execute inference
+
+```
+python -m monai.bundle run evaluating --meta_file configs/metadata.json --config_file configs/inference.json --logging_file configs/logging.conf
```
Comment (wyli) [inline models/wholeBody_ct_segmentation/docs/README.md:139]: Please consider adding the support/documentation of running model inference with CLI on a single file or a folder of image files. Cc @nvahmadi 

Something like:

python -m monai.bundle run evaluating \
    --config_file configs/inference.json \ 
    --datalist "['/test/IBSR_01_ana.nii.gz', '/Users/Documents/MNI152_T1_2mm.nii.gz']"

Comment (nvahmadi) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: Hi, thanks for Cc'ing me, @wyli . I recently discussed multi-image/folder inference using a trained MONAI bundle, but that was in the context of python-based inference, i.e. not CLI. For the CLI, I think the solution is to pass a datalist parameter (as @wyli indicated above), or a dataset.json file. For python-based inference, I have a code snippet that may help. If you're interested @tangy5, please let me know.

Comment (Nic-Ma) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: Hi @wyli @nvahmadi ,

So I think the PR is already good enough?
You can override the `datalist` of the `inference.json` config at runtime CLI:
```json
"datalist": "$list(sorted(glob.glob(@dataset_dir + '/imagesTs/*.nii.gz')))"
```

Thanks.

Comment (nvahmadi) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: Hi @Nic-Ma, thanks, I think this example is good for indicating how to query a folder for `.nii.gz` files. But especially for new users, the documentation would benefit from a concrete example with filepaths, similar to what @wyli indicated (`--datalist "['/test/IBSR_01_ana.nii.gz', '/Users/Documents/MNI152_T1_2mm.nii.gz']"`). Would be good to include that. Other than that, I have no comments w.r.t. the PR.

Comment (tangy5) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: Thanks. I think the idea is good. 

Added the command of the inference with --datalist

Welcome further suggestions on any code snippets. If only doc is needed to change, we can update the PR/bundle.

File: models/wholeBody_ct_segmentation/configs/metadata.json (hunk 1 above)
Comment (tangy5) [reply models/wholeBody_ct_segmentation/configs/metadata.json:14]: Thanks, added.

**Comment by tangy5** [comment]
Hi @yiheng-wang-nv @Nic-Ma ,

All above suggestions are uploaded and addressed. I'm ok to merge it. 
Let me know if there are further comments. Thanks!

**Review by yiheng-wang-nv — APPROVED** [review]

File: models/wholeBody_ct_segmentation/docs/README.md (hunk 2 above)
Comment (wyli) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: I think one of the key features of totalsegmentator is that the command line interface is very easy to use... Shall we polish our CLI inference usage as well? For example, why a user would care about the --meta_file option, how to run it in cpu mode or in --fast mode, how to only output a subset of prediction classes?

Comment (Nic-Ma) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: @wyli Maybe we can add a section in the README doc of this repo to introduce several typical bundle CLI commands?
Or develop another tutorial to show different use cases of CLI commands?
https://github.com/Project-MONAI/tutorials/tree/main/model_zoo
What do you think? CC @yiheng-wang-nv 

Thanks.

Comment (wyli) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: sure, that'll be great if you want to improve the CLI usability for all the bundles. For this particular bundle,  when compared with https://github.com/wasserth/TotalSegmentator#usage  do you think we should also review the CLI usage as well?

Comment (yiheng-wang-nv) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: Hi @wyli @Nic-Ma , thanks for the discussions. I will spend some time to check TotalSegmentator first. As for the potential enhancements of CLI usability, if it's need I will create a separate ticket.

Comment (wyli) [reply models/wholeBody_ct_segmentation/docs/README.md:139]: Thanks, please help create a separate ticket, so that we are aware of the usability gaps.

**Comment by ibro45** [comment]
Hi,

I have two questions:

1) How was the configuration selected? Did you use Auto3DSeg, and if so, with what constraints?

2) There seems to be no test set DICE scores. I compared your 1.5mm model against the original fast (3mm, upsampled to 1.5mm) and slow (1.5mm) nnUNet Totalsegmentator models:
image

Did you also observe a DICE of 0.74 in the case of your network or am I getting something wrong?

**Comment by diazandr3s** [comment]
Hi @ibro45,

Very good questions.

> 1. How was the configuration selected? Did you use Auto3DSeg, and if so, with what constraints?

The whole-body CT segmentation model uses a single network (SegResNet). Both Auto3D and nnUNET train an ensemble of networks with different hyperparameters. This means more computational resources might be needed to train and perform inference when using those two algorithms

> 2. There seems to be no test set DICE scores. I compared your 1.5mm model against the original fast (3mm, upsampled to 1.5mm) and slow (1.5mm) nnUNet Totalsegmentator models:

I wouldn't be surprised if the performance of the whole-body CT segmentation model is slightly worse when compared to nnUNET or Auto3D. Essentially because this model uses a single network instead of an ensemble.

A good exercise would be to compare the Auto3D against the nnUNET results on this dataset.

**Comment by ibro45** [comment]
Thanks @diazandr3s!

I see, I haven't worked with Auto3D yet, but I assumed that it, like nnUNet, configures the network, patch and batch size, etc. based on the dataset. So I was just wondering if that was done first to figure out the "ideal" configuration before training it.

Is there any particular reason why the test scores aren't reported? I can't find any in MONAI VISTA either, where TotalSegmentator was used too.

I'm actually developing a single SegResNet (same architecture as wholebody_ct) for all 104 labels and am getting close to the original 5-model nnUNet TotalSegmentator performance. I was just wondering about the design choices as I'm debating whether I should also adjust the current SegResNet architecture and see if it improves the performance.

**Comment by diazandr3s** [comment]
Hi  @ibro45,

> I see, I haven't worked with Auto3D yet, but I assumed that it, like nnUNet, configures the network, patch and batch size, etc. based on the dataset. So I was just wondering if that was done first to figure out the "ideal" configuration before training it.

That's correct, the Auto3D algorithm configures those hyperparameters. You can learn more about this model here: https://github.com/Project-MONAI/tutorials/tree/main/auto3dseg#performance-benchmarking

For the whole-body model, we used the vanilla SegResNet. No major network hyperparameter search was performed.

> Is there any particular reason why the test scores aren't reported? I can't find any in MONAI VISTA either, where TotalSegmentator was used too.

The initial idea of proposing the whole-body CT segmentation was not to show a better performance than the nnUNET or Auto3D. It was more to demonstrate how the whole body can be segmented using a single network instead of an ensemble. Easier to train and faster inference for downstream tasks.

With regard to the MONAI VISTA, you should hear more details about this model soon. Stay tuned!

> I'm actually developing a single SegResNet (same architecture as wholebody_ct) for all 104 labels and am getting close to the original 5-model nnUNet TotalSegmentator performance. I was just wondering about the design choices as I'm debating whether I should also adjust the current SegResNet architecture and see if it improves the performance.

This is great! Please let us know how that goes. Looking forward to hearing more.
